### DISCUSSION TITLE
i want to split machine code in typescript

### DISCUSSION THREAD

**Discussion by extracold1209** [statelyai/xstate#3643] [Q&A]
I want to divide the machine code of xstate info file units, such as

### combined code

```typescript
const exampleMachine = createMachine({
    id: 'watermarks',
    schema: {
      context: {} as ExampleContexts,
      events: {} as ExampleEvents,
    },
    tsTypes: {} as import('./machine.typegen').Typegen0,
    predictableActionArguments: true,
    initial: 'inactive', // just example. point is split actions
    context: {
      watermarksList: [],
      watermarks: {},
      batchedUpdates: {},
    },
    states: {
      active: {
        on: {
          UPDATE_WATERMARKS: {
            actions: 'batchUpdate',
          },
        },
      },
    },
  }, {
    actions: {
      batchUpdate: assign((context, event) => { /* some logic*/ })
    }
  });
```

### split code (i want)

```typescript

const exampleMachineActions: WhatType =  {
  batchUpdate: assign((context: TContextWantToExampleContext, event: TEventWantToExampleEvents) => { /* some logic*/ }),
}

const exampleMachine = createMachine({
    id: 'watermarks',
    schema: {
      context: {} as ExampleContexts,
      events: {} as ExampleEvents,
    },
    tsTypes: {} as import('./machine.typegen').Typegen0,
    predictableActionArguments: true,
    initial: 'inactive',
    context: {
      watermarksList: [],
      watermarks: {},
      batchedUpdates: {},
    },
    states: {
      active: {
        on: {
          UPDATE_WATERMARKS: {
            actions: 'batchUpdate',
          },
        },
      },
    },
  }, {
    actions: exampleMachineActions, // in this case, exampleMachineActions is can not infer type
  });
```

can i type as i want? it seems that various types written inside xstate cannot be imported for use outside.
in above case, i can use `InternalMachineOptions` instead `WhatType` i think. but this type has no example

it seems to be possible to use `InternalmachineOptions` as follows, but i don't know what type i can put in.

```typescript
const actions: InternalmachineOptions<TContext, TEvent, TResolvedTypesMeta>['actions'] = { /* [key: actionName]: Logic such as assign */ };
```

can i get an example or advice on this? thank you.

(actually, i want to separate as other files all ths values with long logic like services and actions)

**Comment by Andarist** [comment]
Could you share this as a runnable~ example? It could be a codesandbox or a TS layground.

**Comment by extracold1209** [comment]
@Andarist 
ok here is simple runnable code :)

https://codesandbox.io/embed/elated-aj-x12wbq?fontsize=14&hidenavigation=1&theme=dark

and i want to know how to define type heavy and lenghty logics (ex. actions, services in createMachine config)
also in example above, is there no choice to using type guard(ex. `if (event.type !== 'ANYTYPE') return`) to know what type of context, event in `assign` functions's argument is?

**Comment by Andarist** [comment]
You can utilize `MachineOptionsFrom`, like this: https://codesandbox.io/s/silly-liskov-jconn1?file=/src/index.ts:864-925

Thus combined with [typegen](https://xstate.js.org/docs/guides/typescript.html#typegen) should give you very nice types for events given to all implementation types (actions, guards, etc).

**Reply by extracold1209** [reply]
good! it works well :) and when i use this type `MachineOptionsFrom`, i must use withConfig not createMachine's second argument.

additionally, i also found that xstate has more `-From` types. i'll use these types. thx

**Reply by extracold1209** [reply]
hey i got problem when i use `MachineOptionsFrom and withConfig` and `xstate typegen`.

case 1. ⭕️ use typegen when actions in `createMachine`

https://codesandbox.io/s/priceless-lamarr-sssb4b?file=/src/index.ts

in this case autogenerated typegen file has no `missingImplementations`.

case 2. ❌ use typegen when actions not in `createMachine` but define outside using `MachineOptionsFrom`

https://codesandbox.io/s/wonderful-diffie-3xtqw8?file=/src/index.ts

typegen can not detect actions type :(  
but `actions property in missingImplementatinos in Typegen0 interface` change `never` type forceful, then type check pass!

maybe point will be `how to tell typegen that actions has provided using withConfig`

**Reply by extracold1209** [reply]
@Andarist 

could you help me? typegen's feature that is not yet supported or is there something i set up wrong?
